### ISSUE TITLE
APIFilter memory leak?

### ISSUE THREAD

**Issue by fenders** [anotheria/moskito#152]
Hi, 
thanks first of all for the new releases!

We recently updated to MoSKito 2.8.2. Since the update, we notice several millions of 'APISessionImpl's on the heap, which seem to lead to OutOfMemoryErrors. 
The API filter gets registered for all urls within webui's web-fragment [1]. That wasn't the case for webui 2.7.1, where we did not observe that behaviour.

Any idea why that happens?

Could we deactivate the filter or would it break MoSKito?

Thanks and all the best!
Frank

[1]
```
<name>MoSKitoInspectShared</name>
<filter>
	<filter-name>APIFilter</filter-name>
	<filter-class>net.anotheria.anoplass.api.filter.APIFilter</filter-class>
</filter>
<filter-mapping>
	<filter-name>APIFilter</filter-name>
	<url-pattern>/*</url-pattern>
</filter-mapping>
```

ps:
We use MoSKito embedded into our application as described here:
https://confluence.opensource.anotheria.net/display/MSK/Embedding+Moskito-WebUI+Into+Your+Application

**Comment by dvayanu** [comment]
Hi Frank,

this is strange. Can you create a heap dump for analysis? This would be really helpful. If you could create a heap dump, send me a link to it to rosenberg.leon@gmail.com, you don't need to share it with everyone ;.-)

Another question: do you have any tools that access moskito regularly? The only way I can imagine this is happening is if something is hitting protected area over and over again.

regards
Leon

**Comment by fenders** [comment]
Hi Leon, I requested the dump and send it to you as soon as I get it. I am not aware of any tools accessing MoSKito. Best regards, Frank

**Comment by dvayanu** [comment]
I have tried to reproduce it with a latest standalone moskito installation but even after some couple of hundreds hits on the app I've managed to create three objects (h1-h5 are different histograms):
h2.txt:1123:             1             56  net.anotheria.anoplass.api.session.APISessionImpl
h2.txt:1124:             1             56  net.anotheria.anoplass.api.session.APISessionImpl
h3.txt:1340:             1             56  net.anotheria.anoplass.api.session.APISessionImpl
h3.txt:1341:             1             56  net.anotheria.anoplass.api.session.APISessionImpl
h4.txt:1313:             1             56  net.anotheria.anoplass.api.session.APISessionImpl
h4.txt:1314:             1             56  net.anotheria.anoplass.api.session.APISessionImpl
h5.txt: 843:             3            168  net.anotheria.anoplass.api.session.APISessionImpl
h5.txt:1011:             2            112  net.anotheria.anoplass.api.session.APISessionImpl

Are we talking about a web-application and embedded moskito-ui?

**Comment by fenders** [comment]
Yes, it's a web application with an embedded moskito ui. 
The web application itself is getting hit by a large number of REST requests high frequently by different feeders (other applications). 
Here is our moskito specific content/configuration from the applications web.xml:

```
<filter>
	<filter-name>MoskitoUIFilter</filter-name>
	<filter-class>net.anotheria.moskito.webui.MoskitoUIFilter</filter-class>
	<init-param>
		<param-name>path</param-name>
		<param-value>/monitor/</param-value>
	</init-param>
</filter>
<filter-mapping>
	<filter-name>MoskitoUIFilter</filter-name>
	<url-pattern>/monitor/*</url-pattern>
</filter-mapping>
	<!--  adding moskito measurement and command filters -->
<filter>
	<filter-name>RequestURIFilter</filter-name>
	<filter-class>net.anotheria.moskito.web.filters.RequestURIFilter</filter-class>
	<init-param>
		<param-name>limit</param-name>
		<param-value>1500</param-value>
	</init-param>
</filter>
<filter>
	<filter-name>MoskitoCommandFilter</filter-name>
	<filter-class>net.anotheria.moskito.web.filters.MoskitoCommandFilter</filter-class>
</filter>
<filter>
	<filter-name>JourneyFilter</filter-name>
	<filter-class>net.anotheria.moskito.web.filters.JourneyFilter</filter-class>
</filter>
	<filter-mapping>
	<filter-name>RequestURIFilter</filter-name>
	<url-pattern>/*</url-pattern>
</filter-mapping>
<filter-mapping>
	<filter-name>MoskitoCommandFilter</filter-name>
	<url-pattern>/*</url-pattern>
</filter-mapping>
<filter-mapping>
	<filter-name>JourneyFilter</filter-name>
	<url-pattern>/*</url-pattern>
</filter-mapping>

<!-- this listener will init moskito with the context name and cleanup jmx beans and timers on undeploy -->
<listener>
	<listener-class>net.anotheria.moskito.webui.util.StartStopListener</listener-class>
</listener>
<!-- provides information about created sessions -->
<listener>
	<listener-class>net.anotheria.moskito.web.session.SessionCountProducer</listener-class>
</listener>
	<!-- prepared some useful accumulators -->
<listener>
	<listener-class>net.anotheria.moskito.webui.util.SetupPreconfiguredAccumulators</listener-class>
</listener>
```

**Comment by dvayanu** [comment]
Since you only need the APIFilter for the webui (moskito-inspect), I assume changing the mapping to 
```
<filter-mapping>
	<filter-name>APIFilter</filter-name>
	<url-pattern>/monitor/*</url-pattern>
</filter-mapping>
<filter-mapping>
	<filter-name>APIFilter</filter-name>
	<url-pattern>/monitor/</url-pattern>
</filter-mapping>

```should help.

**Comment by dvayanu** [comment]
Hi, did it help?

**Comment by fenders** [comment]
Hi, we deactivated that APIFilter by setting the application to metadata-complete in web.xml. Since we did not observe any issues with using the WebUI afterwards, we did not reactivate the filter and map it to /monitor.
I close the issue. Perhaps a hint in the documentation for using MoSKito embedded would be good. If I understand correctly, everybody who uses it embedded would currently have the same issue?
Thanks for your quick help!
